.github/workflows/test.yaml

```diff
@@ -0,0 +1,80 @@
+name: KBase Handle Service 2 test
+
+on:
+  pull_request:
+    types:
+    - opened
+    - reopened
+    - synchronize
+    - ready_for_review
+  push:
+    # run workflow when merging to main or develop
+    branches:
+      - main
+      - master
+      - develop
+
+jobs:
+  handle_service_tests:
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - python: '3.9.12'
+            mongo: 'mongodb-linux-x86_64-3.6.23'
+            wired_tiger: 'false'
+
+    steps:
+    - name: Check out GitHub repo
+      uses: actions/checkout@v4
+
+    - name: Setup Python
+      uses: actions/setup-python@v5
+      with:
+        python-version: ${{matrix.python}}
+
+    - name: Install dependencies and set up test config
+      shell: bash
+      env:
+        AUTH_SERVICE_URL: https://ci.kbase.us/services/auth/api/legacy/KBase/Sessions/Login
+        AUTH_URL: https://ci.kbase.us/services/auth
+        SHOCK_URL: https://ci.kbase.us/services/shock-api
+        ADMIN_TOKEN: ${{secrets.KBASE_CI_TOKEN}}
+      run: |
+
+        # install python libraries
+        pip install -r requirements.txt  
```

Comment:
If we're going to maintain a requirements list (which is a good idea) we should use pipenv or poetry that handles dependencies properly. Personally I'd go with pipenv since that's what we're using in several other projects (e.g. collections). The Dockerfile should also be updated to install via pipenv. Pipenv also handles the separation between standard and dev dependencies more cleanly than requirements files.

Reply:
Sse pipenv now in GHA. `uwsgi` cannot be directly installed using pip in Docker. So keep it as it is.

Reply:
why not?

Reply:
Failed to build uwsgi: https://github.com/kbase/handle_service2/actions/runs/8993835770/job/24706321954

Reply:
Ok, looking at the dockerfile, I tried to get uwsgi to install with pip as well around January and gave up. IIRC I spent a pretty good chunk of time on it

That being said, there's a couple issues with the current set up:
* The python version used for tests is 3.9 while the version in the dockerfile is 3.8. The python with which we run the prod server needs to be the python we test with.
* We're now specifying requirements both in requirements.txt and in the dockerfile - there needs to be one source of truth
* We're installing mongo in the dockerfile for no reason AFAICT

The best course of action is probably to no longer use the sdkbase images since we no longer need them, but that's a different PR

Reply:
> We're now specifying requirements both in requirements.txt and in the dockerfile - there needs to be one source of truth

We are specifying requirements both in pipenv and in the dockerfile because of uwsgi. What is your recommendation on this? I can remove uwsgi from pipenv and install separately in GHA. Now we have one source of truth, but scattered ...

Reply:
Also, I started with python 3.8.10 (align with the dockerfile), but the version '3.8.10' with architecture 'x64' was not found for Ubuntu 22.04. Therefore, I bumped the version.

Reply:
> Also, I started with python 3.8.10 (align with the dockerfile), but the version '3.8.10' with architecture 'x64' was not found for Ubuntu 22.04 in GHA. 

That's unfortunate. The options I see, in that case, are
* Manually install & use python 3.8 in GHA
* Update the dockerfile to no longer use sdkbase and use a python image. This might fix the uwsgi install problem as well

Reply:
That's one option. If you redo the dockerfile so it doesn't use sdkbase or conda pipenv would probably work

Reply:
why we need to align with 3.8? With this logic, python will never be upgraded in prod.
Prod should align with what works in dev.

Reply:
You don't. The requirement is that we deploy and test with the same version of python

Reply:
Using a python image fix the uwsgi install problem, but now complains `kb-sdk` is missing. 
Any code I can reference on how to correctly install `kb-sdk` inside the dockerfile?

Reply:
What's complaining? You shouldn't need kb-sdk anymore - just check in the compiled files

Reply:
<img width="998" alt="Screenshot 2024-05-08 at 1 23 41 PM" src="https://github.com/kbase/handle_service2/assets/22202096/66b48df2-0859-4881-84f1-828b6e9531fe">

Reply:
Don't run make all, just run the make targets you need. You might not need any, not sure

Reply:
I am not even sure why we still need this dockerfile. It was used for test so we had mongo and test stuff. Now they are moved outside already.

Reply:
The dockerfile is used to build the image in GHA and is what is deployed in Rancher

Reply:
No dockerfile = no service

Reply:
Do we need to run `build-test-script` target in the Dockerfile? Not sure about this one.
We definitely need some of them.

Reply:
Not if you can get the tests to run without it, which I suspect is possible

Reply:
> We're installing mongo in the dockerfile for no reason AFAICT

handler talks to Mango. So we basically assume devops team will take care of it when deploy?

Reply:
The mongo database has to be outside the container, since otherwise everytime the container was recreated all the data would be lost

Reply:
so, dockerfile here just basically starts the server

Reply:
Right

Reply:
why need `dockerize` in the dockerfile?

Reply:
https://github.com/kbase/dockerize

Reply:
👍

Reply:
The Pipfile should separate standard and test dependencies, and only install the standard dependencies in the dockerfile. See Collections for an example

Reply:
Is there a way just installing the standard dependencies in the dockerfile? 
In Collections, it basically install all packages specified in Pipfile.lock. (https://github.com/kbase/collections/blob/827bf2c519cbd5fd7646d4086621ebf8baf2b3a7/Dockerfile#L12)

Reply:
> Is there a way just installing the standard dependencies in the dockerfile?

That's what it does by default. You have to specify --dev to get it to install the develop dependencies as well

Reply:
Here is what I understand:

`pipenv sync`: install dependencies, both standard and development, listed in the lock file.
`pipenv sync --dev`: installs only the development dependencies, listed in the lock file.

If we want to install standard dependencies only in the dockerfile, we probably need to use `--categories` args.

Reply:
```
$ pipenv sync -h
Usage: pipenv sync [OPTIONS]

  Installs all packages specified in Pipfile.lock.

Options:
  --system               System pip management.  [env var: PIPENV_SYSTEM]
  --bare                 Minimal output.
  --categories TEXT
  -d, --dev              Install both develop and default packages  [env var:
                         PIPENV_DEV]
```

Reply:
then examples from Collections are not correct. `--dev` installed both. But in GHA we only want test dependencies while in the Dockerfile default dependencies

Reply:
> then examples from Collections are not correct. 
What examples?
> But in GHA we only want test dependencies 
You need all the dependencies to run tests, otherwise the code under test with throw import errors.

Reply:
fixed 👍

Reply:
:+1: